Use Bayesian updating to compute census summarizations

drewsamnick opened the pull request:
The initial way we were calculating if a school teaches CS was to check if all the census data were consistent and record that value it it was and a **MAYBE** if it wasn't.  Some of the issues with the simple approach:
* There was no good way to make use of the absence of data very well. For example, knowing that a school isn't listed in the AP data set gives us some information, but not enough to completely cancel out other data points.
* There wasn't a good way to use previous years' data since we expect schools to change over time.
* There was no good way to weight the different data points relative to each other. That means that a single disagreement would automatically flip the classification to maybe. We want something more nuanced than that.

Writing out a collection of rules for how to combine the different types of data is very unwieldy since there are so many permutations. To avoid that, we're going to try out a Bayesian approach. That allows us to specify just the likelihood of seeing each type of data for schools that do and do not teach CS and then combine all the data together into a measurement of how strongly we believe that the school teaches CS. 

I tried to document the approach in the code and I am especially interested in whether that explanation needs more  detail.

drewsamnick commented:
@aoby As somebody with less stats experience, would you feel comfortable touching this code (for example, adding in a section to process a new flavor of census data)? If not, where would you want to see me detailed explanation?

aoby commented:
Yes, I think so.

It does seem like this pattern is repeated:
```ruby
likelihoods[:teaches].push CATEGORY[:given_teaches][:field]
likelihoods[:not_teaches].push CATEGORY[:given_does_not_teach][:field]
```
It might be easier to digest and extend if that were condensed, e.g. 
```ruby
adjust_likelihood_for(CATEGORY, field)
```
?

Since there are only a few of these it's not so bad as it stands.

ryansloan commented:
Probably better to have @bencodeorg take a look than me. My Ruby-fu is weak but I wonder a bit about some of the priors. We have done some educated guesswork since we don't have labeled data. Is there any kind of validation we can do on these rates? (Things like AP school code mapping accuracy, or the likelihood that a school does/does not teach given its presence in a state data set)

What does the distribution and audit of classifications look like when you run this model?

ryansloan commented on dashboard/app/models/census/census_summary.rb:
```diff
@@ -168,18 +295,43 @@ def self.summarize_census_data
+          # Assuming some chance that the school will stop teaching we degrade the
+          # old belief.
+          # Use an uninformed prior (0.5) if we don't have a previous year's belief.
+          prior = previous_years_belief ? (previous_years_belief * 0.9) : 0.5
```
I'm not sure what's standard for this type of model, but given that (I assume) the base rate of teaching CS is low as a percent of all schools, is .5 a reasonable uninformed prior to use? (Seems high)

drewsamnick replied:
0.5 is the *uninformed* prior because it doesn't take any outside information into account and just says that without any data we have no preference for thinking that the school does or does not teach CS. A more informed value could be something around 0.4 since that is the number we are currently reporting. In practice that doesn't make much difference to the posteriors once we have some data about a school.

ryansloan commented on dashboard/app/models/census/census_summary.rb:
```diff
@@ -168,18 +295,43 @@ def self.summarize_census_data
+            # P(not teaches | data) = P(not teaches) * P(data_1 | not teaches) * P(data_2 | not teaches) * ... / P(data)
+            #
+            # P(teaches) is the prior, P(not teaches) is 1 - prior
+            # P(data_n | teaches) and P(data_n | no0t teaches) are the likelihoods for data_n
```
typ0 in the c0mment 😉

bencodeorg commented:
Probabilistic thinking at Code.org?? 😍 Haven't reviewed details, but agree with Ryan's questions re: priors. Also, thoughts on how you'd report this in aggregate? How would we validate different thresholds for counting as "teaching" (i.e., is there some gold standard dataset you'd use for evaluation?)

drewsamnick commented:
@ryansloan The values used for the likelihoods here are just a first pass. We're still debating exactly which method to use to summarize the census data. If we end up going further with this approach then we'll come back and refine those. Unfortunately there isn't much we can do to validate those rates. These will be subjective values out of necessity. The alternative approach is to have a matrix of potential combinations of data we might see for a school and to hard code rules about which data sources 'win' over others. That is at least as subjective in my opinion. 

I won't have details on how this works until the code is in production. At that point we will compare it with the simple approach we're using right now (count as a yes if all data is YES, count as a no if all data says NO, count as a maybe is the data conflict.)

@bencodeorg Right now for validation we are just planning to do some spot checking to make sure the results are not totally nonsensical. Reporting in the aggregate will be based on the classification of `teaches_cs` as YES, NO, MAYBE, or NULL (no data) for each school.

bencodeorg commented:
I see, was just wondering if you were planning on playing with thresholds and seeing how that affected results. Would be cool to have a few hundred schools hand tagged as a test set to see what threshold aligns most closely with what a human would deem likely "teaching".

bencodeorg commented:
Keep me in the loop on this, as I think I'd probably be involved in maintenance after you're gone.

drewsamnick commented:
@bencodeorg Yes, I'll keep you in the loop. 
I like your suggestion of having a validation set to sanity check any approaches we consider. There are plenty of schools with conflicting data in production so I can mine that data for examples and create some test cases.

drewsamnick commented:
I refactored the code a little bit and added a bunch of test cases. The tests cases validate that we get expected classifications (YES, NO, MAYBE) and not the specifics of this methodology. They are reusable even if we ultimately decide not to go with the Bayesian approach.
